crates/bevy_asset/src/ser.rs

```diff
@@ -75,31 +75,14 @@ impl<'de, T: Asset> Deserialize<'de> for Handle<T> {
 }
 
 impl AssetServer {
-    pub fn serialize_with_asset_refs<S, T>(
-        &self,
-        serializer: S,
-        value: &T,
-    ) -> Result<S::Ok, S::Error>
+    /// Enables asset references to be serialized or deserialized
+    pub fn with_asset_refs_serialization<F, T>(&self, f: F) -> T
     where
-        S: Serializer,
-        T: Serialize,
-    {
-        ASSET_SERVER.with(|key| {
-            key.replace(Some(self.clone()));
-            let result = value.serialize(serializer);
-            key.replace(None);
-            result
-        })
-    }
-
-    pub fn deserialize_with_asset_refs<'de, D, T>(&self, deserializer: D) -> Result<T, D::Error>
-    where
-        D: Deserializer<'de>,
-        T: Deserialize<'de>,
+        F: FnOnce() -> T,
     {
         ASSET_SERVER.with(|key| {
             key.replace(Some(self.clone()));
-            let result = T::deserialize(deserializer);
+            let result = (f)();
```

Comment:
Do you have an example use case of how this API is better than the previous one?
It worries me here that `T` is not constrained on `Serialize` or `Deserialize`

Reply:
the last API didn't allow to `de::DeserializeSeed` to be used for stateful de-serialization

Reply:
Yep, that makes sense.
However, I'm still not a big fan of just having the parameter be `f: impl FnOnce() -> T`. 
you could give this any `f` that returns any type, which as a user of the function, isn't very informative.
If we could come up with some way to constrain `T` to better express that `T` needs to be serializable/deserializable via the `AssetServer` context, that would be best IMO.